build.wake

```diff
@@ -0,0 +1,430 @@
+#
+# Utility Functions
+#
+
+# This expects a sorted list of paths.
+# If a path is a prefix of the path that follows it, it is filtered out.
+def filterDirs sortedFiles =
+  def helper acc list = match list
+    head1, head2, tail =
+      def isPrefixOf path = matches (regExpCat (path.quote, `(/.+)?`, Nil))
+      if isPrefixOf head1 head2
+      then helper acc (head2, tail)
+      else helper (head1, acc) (head2, tail)
+    _ = list ++ acc
+  helper Nil sortedFiles
+  | reverse
+
+def getDestDirs srcDir destDir inputs =
+  def srcDirRegex = regExpCat (`^`, "{srcDir}/..".simplify.quote, `/+`, Nil)
+  inputs
+  | mapFlat (
+    _
+    | getPathName
+    | simplify
+    | replace srcDirRegex ""
+    | extract `(.*/)[^/]*`
+    | map ("{destDir}/{_}")
+  )
+  | distinctBy scmp
+
+def copyDir srcDir destDir fileInputs =
+  def dirInputs =
+    getDestDirs srcDir destDir fileInputs
+    | map mkdir
+  def inputs = dirInputs ++ fileInputs
+  def from_ = simplify "{srcDir}/.."
+  def srcs =
+    fileInputs
+    | map (
+      _
+      | getPathName
+      | from_.relative
+      | simplify
+    )
+    | filter (! matches `\.\./.*` _)
+    | sortBy (_ <~ _)
+    | filterDirs
+  def cmdline =
+    "rsync", "--inplace", "--relative",
+    "--ignore-times", srcs ++ (relative from_ destDir, Nil)
+  makePlan cmdline inputs
+  | setPlanDirectory from_
+  | runJob
+  | getJobOutputs
+
+def filterWakeAndDotFiles =
+  _
+  | filter (!matches `(.*/)*\..*` _.getPathName)
+  | filter (!matches `.*\.wake` _.getPathName)
+
+#
+# Plan Resources
+#
+
+global def defaultFreedomESDKResources =
+  "riscv-tools/2019.08.0",
+  "python/python/3.7.1",
+  Nil
+
+#
+# Virtualenv Creation
+#
+# Freedom E SDK depends on a number of Python libraries. Here we enable
+# the installation of a single, workspace-wide virtualenv which can be
+# preinstalled using either the `preinstallFreedomESDKVirtualenv` target
+# or the global `preinstall` topic. If the virtualenv has not been
+# preinstalled, it will be created on-demand when needed during the
+# creation of a `FreedomESDKTarget`.
+#
+
+global def freedomESDKBuildDir = "build/freedom-e-sdk"
+global def freedomESDKVenvDir = "{freedomESDKBuildDir}/venv"
```

Comment:
Similar to above, I think the top-level projects will want to have more control over where the build files are created. With this approach, there's a bit of a namespacing issue, where we have to come up with unique build directory names. Although naming it after the repository like this more or less prevents conflicts, this may not necessarily present an organization that makes sense to the user of a top-level project.

For example, if I'm running through the Chisel -> Verilog -> SDK -> RTL simulation flow, such as in the [block-pio-sifive](https://github.com/sifive/block-pio-sifive) flow, the user of the system really cares foremost about the primary artifacts produced in each step above (design Verilog files, design-specific SDK, RTL simulator), and not stuff like "here are the .jar files created by each of the Chisel dependencies, each in its own directory directly in `build/`.

I think allowing these to be overridable by the caller will provide greater flexibility for top-level projects to organize their build directories in a way that makes the most sense for that project.